packages/ember-data/lib/serializers/json-serializer.js

```diff
@@ -274,11 +276,42 @@ export default Serializer.extend({
   normalizeErrors: function(typeClass, hash) {
     this.normalizeId(hash);
     this.normalizeAttributes(typeClass, hash);
-    this.normalizeRelationships(typeClass, hash);
+    this.normalizeRelationshipErrors(typeClass, hash);
     this.normalizeUsingDeclaredMapping(typeClass, hash);
   },
 
   /**
+    @method normalizeRelationshipErrors
+    @private
+  */
+  normalizeRelationshipErrors: function(typeClass, hash) {
+    if (!this.keyForRelationship) {
+      return;
+    }
+
+    typeClass.eachRelationship(function(key, relationship) {
+      var payloadKey = this.keyForRelationship(key, relationship.kind, 'deserialize');
+      if (key !== payloadKey && hash.hasOwnProperty(payloadKey)) {
+        hash[key] = hash[payloadKey];
+        delete hash[payloadKey];
+      }
+
+      if (!hash || !hash[key]) {
+        return;
+      }
+      if (relationship.kind === 'hasMany' && isArray(hash[key])) {
+        forEach.call(hash[key], function(errors) {
+          if (errors) {
+            this.normalizeErrors(relationship.type, errors);
```

Comment:
>Also, I think we should add an assertion/warning for the case when the length of the errors array is not the same length of the array of records [...]

While I agree with the idea, I don't think it's the serializer role to do that—and we don't have an instance of the record there— it would be at the record level in the current implementation (`adapterDidInvalidate`), which I feel is out of the scope of this PR.